File: bcs-services/bcs-webconsole/route/permissions.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

package route

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/pkg/errors"

	"github.com/Tencent/bk-bcs/bcs-services/bcs-webconsole/console/components/bcs"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-webconsole/console/components/iam"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-webconsole/console/config"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-webconsole/console/i18n"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-webconsole/console/sessions"
	"github.com/Tencent/bk-bcs/bcs-services/bcs-webconsole/console/types"
)

// PermissionRequired 权限控制，必须都为真才可以
func PermissionRequired() gin.HandlerFunc {
	return func(c *gin.Context) {
		if c.Request.Method == http.MethodOptions {
			c.Next()
			return
		}

		authCtx := MustGetAuthContext(c)

		// 校验项目，集群信息的正确性
		if err := ValidateProjectCluster(c, authCtx); err != nil {
			c.AbortWithStatusJSON(http.StatusBadRequest, types.APIResponse{
				Code:      types.ApiErrorCode,
				Message:   err.Error(),
				RequestID: authCtx.RequestId,
			})
			return
		}

		c.Set("auth_context", authCtx)

		// 管理员不校验权限, 包含管理员凭证
		if config.G.IsManager(authCtx.Username, authCtx.ClusterId) {
			c.Next()
			return
		}

		if err := initContextWithIAMProject(c, authCtx); err != nil {
			c.AbortWithStatusJSON(http.StatusForbidden, types.APIResponse{
				Code:      types.ApiErrorCode,
				Message:   err.Error(),
				RequestID: authCtx.RequestId,
			})
			return
		}

		c.Next()
	}
}

// SessionRequired session 权限校验
func SessionRequired() gin.HandlerFunc {
	return func(c *gin.Context) {
		if c.Request.Method == http.MethodOptions {
			c.Next()
			return
		}

		authCtx := MustGetAuthContext(c)

		podCtx, err := validateSession(c, authCtx)
		if err != nil {
			c.AbortWithStatusJSON(http.StatusForbidden, types.APIResponse{
				Code:      types.ApiErrorCode,
				Message:   err.Error(),
				RequestID: authCtx.RequestId,
			})
			return
		}

		authCtx.BindSession = podCtx
		c.Set("auth_context", authCtx)

		c.Next()
	}
}

// 校验用户session权限
func validateSession(c *gin.Context, authCtx *AuthContext) (*types.PodContext, error) {
	sessionId := GetSessionId(c)
	if sessionId == "" {
		return nil, errors.New("session_id is required")
	}

	podCtx, err := sessions.NewStore().OpenAPIScope().Get(c.Request.Context(), sessionId)
	if err != nil {
		return nil, errors.New("session已经过期或不合法")
	}

	if config.G.IsManager(authCtx.Username, podCtx.ClusterId) {
		return podCtx, nil
	}

	if podCtx.HasPerm(authCtx.Username) {
		return podCtx, nil
	}

	return nil, errors.New("用户无权限登入此session")
}

// ValidateProjectCluster xxx
func ValidateProjectCluster(c *gin.Context, authCtx *AuthContext) error {
	projectId := GetProjectIdOrCode(c)
	if projectId == "" {
		return errors.New("project_id or code is required")
	}

	clusterId := GetClusterId(c)
	if clusterId == "" {
		return errors.New("clusterId required")
	}

	project, err := bcs.GetProject(c.Request.Context(), config.G.BCS, projectId)
	if err != nil {
		return errors.Wrap(err, i18n.T(c, "项目不正确"))
	}

	cluster, err := bcs.GetCluster(c.Request.Context(), project.ProjectId, clusterId)
	if err != nil {
		return errors.Wrap(err, i18n.T(c, "项目或者集群Id不正确"))
	}

	authCtx.BindProject = project
	authCtx.ProjectId = project.ProjectId
	authCtx.ProjectCode = project.Code
	authCtx.TenantId = project.TenantID

	authCtx.BindCluster = cluster
	authCtx.ClusterId = cluster.ClusterId

	return nil
}

// initContextWithIAMProject Dev环境, 可以设置环境变量
func initContextWithIAMProject(c *gin.Context, authCtx *AuthContext) error {
	if config.G.Auth.Disable {
		return nil
	}

	namespace := GetNamespace(c)
	allow, err := iam.IsAllowedWithResource(c.Request.Context(), authCtx.ProjectId, authCtx.ClusterId, namespace,
		authCtx.Username, authCtx.TenantId)
	if err != nil {
		return err
	}
	if !allow {
		return errors.New(i18n.T(c, "没有权限"))
	}

	return nil
}

// CredentialRequired xxx
func CredentialRequired() gin.HandlerFunc {
	return func(c *gin.Context) {
		if c.Request.Method == http.MethodOptions {
			c.Next()
			return
		}
		authCtx := MustGetAuthContext(c)

		if err := ValidateProjectCluster(c, authCtx); err != nil {
			c.AbortWithStatusJSON(http.StatusForbidden, types.APIResponse{
				Code:      types.ApiErrorCode,
				Message:   err.Error(),
				RequestID: authCtx.RequestId,
			})
			return
		}

		c.Set("auth_context", authCtx)

		bkAppCode := authCtx.BKAppCode()

		if bkAppCode == "" {
			c.AbortWithStatusJSON(http.StatusForbidden, types.APIResponse{
				Code:      types.ApiErrorCode,
				Message:   "not valid bk apigw request",
				RequestID: authCtx.RequestId,
			})
			return
		}

		// namespace 模式
		namespace := getNamespace(getRequestBody(c.Request))

		// 校验项目 and 集群 and 命名空间权限，
		// 其中共享集群命名空间校验，只支持 namespace/pod_name/container_name 模式，container_id 模式在共享集群检索效率较低，暂不支持
		scopeValues := map[config.ScopeType]string{
			config.ScopeProjectCode: authCtx.ProjectCode,
			config.ScopeClusterId:   authCtx.ClusterId,
			config.ScopeNamespace:   namespace,
		}

		switch {
		case config.G.ValidateCred(config.CredentialAppCode, bkAppCode, scopeValues):
		default:
			msg := fmt.Sprintf(
				"app %s have no permission, %s, %s, namespace<%s>", bkAppCode, authCtx.BindProject, authCtx.BindCluster, namespace)
			c.AbortWithStatusJSON(http.StatusForbidden, types.APIResponse{
				Code:      types.ApiErrorCode,
				Message:   msg,
				RequestID: authCtx.RequestId,
			})
			return
		}

		c.Next()
	}
}

// ManagersRequired 校验用户是否为manager
func ManagersRequired() gin.HandlerFunc {
	return func(c *gin.Context) {
		authCtx := MustGetAuthContext(c)
		user := authCtx.Username
		managers := config.G.Base.Managers
		for _, manager := range managers {
			if user == manager {
				c.Next()
				return
			}
		}
		c.AbortWithStatusJSON(http.StatusForbidden, types.APIResponse{
			Code:      types.ApiErrorCode,
			Message:   fmt.Sprintf("%s is not managers, has no permission", user),
			RequestID: authCtx.RequestId,
		})
	}
}

// 获取请求体
func getRequestBody(r *http.Request) []byte {
	// 读取请求体
	body, _ := io.ReadAll(r.Body)
	// 恢复请求体
	r.Body = io.NopCloser(bytes.NewBuffer(body))
	return body
}

// 其他内容忽略
type namespaceOnly struct {
	Namespace string `json:"namespace"`
}

// 获取body里面的namespace
func getNamespace(b []byte) string {
	namespace := namespaceOnly{}
	_ = json.Unmarshal(b, &namespace)
	return namespace.Namespace
}

File: bcs-ui/backend/utils/permissions.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
from rest_framework.permissions import BasePermission

from backend.components import paas_cc
from backend.iam.permissions.resources import ProjectPermCtx, ProjectPermission
from backend.utils import FancyDict
from backend.utils.cache import region
from backend.utils.error_codes import error_codes

# 跳过路径为projects、projects_pub的namespace的项目校验
SKIP_REQUEST_NAMESPACE = ["projects", "projects_pub"]


class HasProject(BasePermission):
    def has_permission(self, request, view):
        project_id = view.kwargs.get("project_id")
        if not project_id:
            return True

        if request.user.is_superuser:
            return True

        perm_ctx = ProjectPermCtx(username=request.user.username, project_id=project_id)
        return ProjectPermission().can_view(perm_ctx, raise_exception=False)


class HasIAMProject(BasePermission):
    def has_permission(self, request, view):
        project_id = view.kwargs.get("project_id")
        if not project_id:
            return True

        if request.user.is_superuser:
            return True

        access_token = request.user.token.access_token

        project_code = self.get_project_code(access_token, project_id)
        if not project_code:
            return False

        perm_ctx = ProjectPermCtx(
            username=request.user.username, project_id=self.get_project_id(access_token, project_id)
        )
        return ProjectPermission().can_view(perm_ctx, raise_exception=False)

    def get_project_code(self, access_token, project_id):
        """获取project_code
        缓存较长时间
        """
        cache_key = f"BK_DEVOPS_BCS:PROJECT_CODE:{project_id}"
        project_code = region.get(cache_key, expiration_time=3600 * 24 * 30)
        if not project_code:
            # 这里的project_id对应实际的project_id或project_code, paas_cc接口兼容了两种类型的查询
            result = paas_cc.get_project(access_token, project_id)
            if result.get("code") != 0:
                return None
            project_code = result["data"]["english_name"]
            region.set(cache_key, project_code)
        return project_code

    def get_project_id(self, access_token, project_id):
        """获取project_id
        缓存较长时间
        # TODO 临时使用
        """
        cache_key = f"BK_DEVOPS_BCS:REAL_PROJECT_ID:{project_id}"
        real_project_id = region.get(cache_key, expiration_time=3600 * 24 * 30)
        if not real_project_id:
            # 这里的project_id对应实际的project_id或project_code, paas_cc接口兼容了两种类型的查询
            result = paas_cc.get_project(access_token, project_id)
            if result.get("code") != 0:
                return None
            real_project_id = result["data"]["project_id"]
            region.set(cache_key, real_project_id)
        return real_project_id


class ProjectHasBCS(BasePermission):
    def has_permission(self, request, view):
        project_id = view.kwargs.get("project_id")
        if not project_id:
            return True

        access_token = request.user.token.access_token

        request_namespace = request.resolver_match.namespace
        project = self.has_bcs_service(access_token, project_id, request_namespace)

        # 赋值给request.project
        project["project_code"] = project["english_name"]
        request.project = project

        return True

    def has_bcs_service(self, access_token, project_id, request_namespace):
        """判断是否开启容器服务
        开启后就不能关闭，所以缓存很久，默认30天
        """
        cache_key = f"BK_DEVOPS_BCS:HAS_BCS_SERVICE:{project_id}"
        project = region.get(cache_key, expiration_time=3600 * 24 * 30)

        if not project or not isinstance(project, FancyDict):
            result = paas_cc.get_project(access_token, project_id)
            project = result.get("data") or {}

            # coes: container orchestration engines
            project['coes'] = project['kind']
            try:
                from backend.container_service.projects.utils import get_project_kind

                # k8s类型包含kind为1(bcs k8s)或其它属于k8s的编排引擎
                project['kind'] = get_project_kind(project['kind'])
            except ImportError:
                pass

            project = FancyDict(project)

            if request_namespace in SKIP_REQUEST_NAMESPACE:
                # 如果是SKIP_REQUEST_NAMESPACE，有更新接口，不判断kind
                if project.get("cc_app_id") != 0:
                    region.set(cache_key, project)

            elif project.get("cc_app_id") != 0:
                region.set(cache_key, project)
            else:
                # 其他抛出没有开启容器服务
                raise error_codes.NoBCSService()

        return project
